lib/nxdk/format.c

```diff
@@ -0,0 +1,148 @@
+#include <nxdk/format.h>
+#include <nxdk/fatx.h>
+
+#include <stdlib.h>
+#include <windows.h>
+#include <xboxkrnl/xboxkrnl.h>
+
+bool nxFormatVolume (const char *volumePath, uint32_t bytesPerCluster)
+{
+    ANSI_STRING volumePathStr;
+    OBJECT_ATTRIBUTES objattr;
+    IO_STATUS_BLOCK ioStatus;
+    HANDLE handle;
+    NTSTATUS status;
+
+    // Default cluster size is 16KiB (32 sectors)
+    if (bytesPerCluster == 0) {
+        bytesPerCluster = 16384;
+    }
+
+    RtlInitAnsiString(&volumePathStr, volumePath);
+    InitializeObjectAttributes(&objattr, &volumePathStr, OBJ_CASE_INSENSITIVE, NULL, NULL);
+
+    status = NtOpenFile(&handle, SYNCHRONIZE | FILE_READ_DATA | FILE_WRITE_DATA, &objattr, &ioStatus, 0, FILE_SYNCHRONOUS_IO_NONALERT | FILE_NO_INTERMEDIATE_BUFFERING);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return false;
+    }
+
+    // Get geometry of the entire disk, mainly to get the sector size
+    DISK_GEOMETRY diskGeometry;
+    status = NtDeviceIoControlFile(handle, NULL, NULL, NULL, &ioStatus, IOCTL_DISK_GET_DRIVE_GEOMETRY, NULL, 0, &diskGeometry, sizeof(diskGeometry));
+    if (!NT_SUCCESS(status)) {
+        NtClose(handle);
+        SetLastError(RtlNtStatusToDosError(status));
+        return false;
+    }
+
+    // Get information about the partition we format, for size checks
+    PARTITION_INFORMATION partInfo;
+    status = NtDeviceIoControlFile(handle, NULL, NULL, NULL, &ioStatus, IOCTL_DISK_GET_PARTITION_INFO, NULL, 0, &partInfo, sizeof(partInfo));
+    if (!NT_SUCCESS(status)) {
+        NtClose(handle);
+        SetLastError(RtlNtStatusToDosError(status));
+        return false;
+    }
+
+    uint32_t alignment = ROUND_TO_PAGES(diskGeometry.BytesPerSector);
+    if (bytesPerCluster < alignment) {
+        bytesPerCluster = alignment;
+    }
+
+    uint32_t superblockBytes = PAGE_SIZE;
+    if (superblockBytes < alignment) {
+        superblockBytes = alignment;
+    }
+
+    uint32_t numclusters = partInfo.PartitionLength.QuadPart / bytesPerCluster + 1;
+    uint32_t clusterTableSize;
```

Comment:
You always refer to 'clusterTable', might want to think about renaming it to FAT, as that's more relatable. (Thats an argument about taste tho)

Reply:
The term FAT feels a bit too ambiguous to me, since it's also often used to refer to the whole file system, and I also want to avoid the impression that this is regular FAT - a lot of the on-disk format is the same, but the fs is not identical (for example the upper bound for FAT16 according to MS's spec is 65525 clusters, but for FATX16 it seems to be 65520, judging from the code I RE'd).